serial_x.go

```diff
@@ -0,0 +1,116 @@
+// +build !windows
+
+package serial
+
+import (
+	"time"
+	"io/ioutil"
+	"regexp"
+	"strings"
+)
+
+const (
+	devFolder = "/dev"
+	regexFilter = `^(cu|tty)\.?.*`
+)
+
+/*
+This function was taken with minor modifications from the go.bug.st/serial package (https://github.com/bugst/go-serial), and is subject to the conditions of its license (reproduced below):
+
+Copyright (c) 2014, Cristian Maglie.
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+   notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+   notice, this list of conditions and the following disclaimer in
+   the documentation and/or other materials provided with the
+   distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+   contributors may be used to endorse or promote products derived
+   from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+*/
+func GetPortsList() ([]string, error) {
+	files, err := ioutil.ReadDir(devFolder)
+	if err != nil {
+		return nil, err
+	}
+
+	ports := make([]string, 0, len(files))
+	for _, f := range files {
+		// Skip folders
+		if f.IsDir() {
+			continue
+		}
+
+		// Keep only devices with the correct name
+		match, err := regexp.MatchString(regexFilter, f.Name())
+		if err != nil {
+			return nil, err
+		}
+		if !match {
+			continue
+		}
+
+		portName := devFolder + "/" + f.Name()
+
+		// Check if serial port is real or is a placeholder serial port "ttySxx"
+		if strings.HasPrefix(f.Name(), "ttyS") {
+			port, err := openPort(portName, 9600, 100)
```

Comment:
This seems pretty ad-hoc.

I'd really like to see something based on syscalls, a C-api, /proc filesystem, character device attributes, etc.

Listing by filename does not seem robust or good enough to make part of the API.

Something like this looks a little better, but more complicated, and still appears to be full of hacks:
https://github.com/openSUSE/hwinfo/blob/9ddd7c1a771e88d47fe986443943ac21cf538945/src/hd/serial.c#L163

I'm not sure it's worth implementing.

Can you describe the use case you have for GetPortsList()?  I have never had a use for a list of all of the serial ports on the system.  I suppose if you were making a GUI that required a serial port, you could present the user with a list of possible serial ports?